moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1169 @@
+            NUMBER_CHAR_NONE -> {
+              negative = true
+              last = NUMBER_CHAR_SIGN
+              i++
+              continue
+            }
+            NUMBER_CHAR_EXP_E -> {
+              last = NUMBER_CHAR_EXP_SIGN
+              i++
+              continue
+            }
+          }
+          return PEEKED_NONE
+        }
+        '+' -> {
+          if (last == NUMBER_CHAR_EXP_E) {
+            last = NUMBER_CHAR_EXP_SIGN
+            i++
+            continue
+          }
+          return PEEKED_NONE
+        }
+        'e', 'E' -> {
+          if (last == NUMBER_CHAR_DIGIT || last == NUMBER_CHAR_FRACTION_DIGIT) {
+            last = NUMBER_CHAR_EXP_E
+            i++
+            continue
+          }
+          return PEEKED_NONE
+        }
+        '.' -> {
+          if (last == NUMBER_CHAR_DIGIT) {
+            last = NUMBER_CHAR_DECIMAL
+            i++
+            continue
+          }
+          return PEEKED_NONE
+        }
+        else -> {
+          if (c !in '0'..'9') {
+            if (!isLiteral(c.code)) break@charactersOfNumber
+            return PEEKED_NONE
+          }
+          when (last) {
+            NUMBER_CHAR_SIGN, NUMBER_CHAR_NONE -> {
+              value = -(c - '0').toLong()
+              last = NUMBER_CHAR_DIGIT
+            }
+            NUMBER_CHAR_DIGIT -> {
+              if (value == 0L) {
+                return PEEKED_NONE // Leading '0' prefix is not allowed (since it could be octal).
+              }
+              val newValue = value * 10 - (c - '0').toLong()
+              fitsInLong = fitsInLong and
+                (
+                  value > MIN_INCOMPLETE_INTEGER ||
+                    value == MIN_INCOMPLETE_INTEGER &&
+                    newValue < value
+                  )
+              value = newValue
+            }
+            NUMBER_CHAR_DECIMAL -> last = NUMBER_CHAR_FRACTION_DIGIT
+            NUMBER_CHAR_EXP_E, NUMBER_CHAR_EXP_SIGN -> last = NUMBER_CHAR_EXP_DIGIT
+          }
+        }
+      }
+      i++
+    }
+
+    // We've read a complete number. Decide if it's a PEEKED_LONG or a PEEKED_NUMBER.
+    return when {
+      last == NUMBER_CHAR_DIGIT &&
+        fitsInLong &&
+        (value != Long.MIN_VALUE || negative)
+        && (value != 0L || !negative) -> {
+        peekedLong = if (negative) value else -value
+        buffer.skip(i)
+        PEEKED_LONG.also { peeked = it }
+      }
+      last == NUMBER_CHAR_DIGIT ||
+        last == NUMBER_CHAR_FRACTION_DIGIT
+        || last == NUMBER_CHAR_EXP_DIGIT -> {
+        peekedNumberLength = i.toInt()
+        PEEKED_NUMBER.also { peeked = it }
+      }
+      else -> PEEKED_NONE
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun isLiteral(c: Int): Boolean {
+    return when (c.toChar()) {
+      '/', '\\', ';', '#', '=' -> {
+        checkLenient() // fall-through
+        false
+      }
+      '{', '}', '[', ']', ':', ',', ' ', '\t', '\u000C'/*\f*/, '\r', '\n' -> false
```

Comment:
> `'\u000C'`

Another fun fact, Kotlin does not recognize `\f` as a valid escape; this was my solution -- although obviously open to other options.